dkg/dkg.go

```diff
@@ -48,6 +56,86 @@ type Config struct {
 	TestSigning bool
 }
 
+// sigData includes the fields obtained from sigdb when threshold is reached.
+type sigData struct {
+	duty   core.DutyType
+	pubkey core.PubKey
+	psigs  []core.ParSignedData
+}
+
+// exchanger is responsible for exchanging partial signatures between peers on libp2p.
+type exchanger struct {
+	sigChan chan sigData
+	sigex   *parsigex.ParSigEx
+	sigdb   *parsigdb.MemDB
+	numVals int
+}
+
+func newExchanger(tcpNode host.Host, peerIdx int, peers []peer.ID, vals int) *exchanger {
+	ex := &exchanger{
+		// threshold is len(peers) to wait until we get all the partial sigs from all the peers per DV
+		sigdb:   parsigdb.NewMemDB(len(peers)),
+		sigex:   parsigex.NewParSigEx(tcpNode, peerIdx, peers),
+		sigChan: make(chan sigData),
+		numVals: vals,
+	}
+
+	// Wiring core workflow components
+	ex.sigdb.SubscribeInternal(ex.sigex.Broadcast)
+	ex.sigdb.SubscribeThreshold(ex.pushPsigs)
+	ex.sigex.Subscribe(ex.sigdb.StoreExternal)
+
+	return ex
+}
+
+// exchange exhanges partial signatures of lockhash/deposit-data among dkg participants and returns all the partial
+// signatures of the group according to public key of each DV.
+func (e *exchanger) exchange(ctx context.Context, duty core.DutyType, set core.ParSignedDataSet) (map[core.PubKey][]core.ParSignedData, error) {
+	// Start the process by storing current peer's ParSignedDataSet
+	err := e.sigdb.StoreInternal(ctx, core.Duty{Type: duty}, set)
+	if err != nil {
+		return nil, err
+	}
+
+	sets := make(map[core.PubKey][]core.ParSignedData)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case peerSet := <-e.sigChan:
+			if duty != peerSet.duty {
+				// Do nothing if duty doesn't match
+				continue
+			}
+
+			switch peerSet.duty {
+			case dutyLock, dutyDepositData:
```

Comment:
you already checked duty above, so no need to switch again, can remove the switch completely